fast_lio_sam_qn/include/fast_lio_sam_qn.h

```diff
@@ -0,0 +1,118 @@
+///// GTSAM
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/slam/BetweenFactor.h>
+#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
+#include <gtsam/nonlinear/Values.h>
+#include <gtsam/nonlinear/ISAM2.h>
+///// coded headers
+#include "loop_closure.h"
+#include "pose_pcd.hpp"
+#include "utilities.hpp"
+
+namespace fs = std::filesystem;
+using namespace std::chrono;
+typedef message_filters::sync_policies::ApproximateTime<nav_msgs::Odometry, sensor_msgs::PointCloud2> odom_pcd_sync_pol;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+class FastLioSamQn
+{
+  private:
+    ///// basic params
+    std::string map_frame_;
+    std::string package_path_;
+    std::string seq_name_;
+    ///// shared data - odom and pcd
+    std::mutex realtime_pose_mutex_, keyframes_mutex_;
+    std::mutex graph_mutex_, vis_mutex_;
+    Eigen::Matrix4d last_corrected_pose_ = Eigen::Matrix4d::Identity();
+    Eigen::Matrix4d odom_delta_ = Eigen::Matrix4d::Identity();
+    PosePcd current_frame_;
+    std::vector<PosePcd> keyframes_;
+    int current_keyframe_idx_ = 0;
+    bool is_initialized_ = false;
+    ///// graph and values
+    std::shared_ptr<gtsam::ISAM2> isam_handler_ = nullptr;
+    gtsam::NonlinearFactorGraph gtsam_graph_;
+    gtsam::Values init_esti_;
+    gtsam::Values corrected_esti_;
+    double keyframe_thr_;
+    double voxel_res_;
+    int sub_key_num_;
+    std::vector<std::pair<size_t, size_t>> loop_idx_pairs_; //for vis
+    bool loop_added_flag_ = false; //for opt
+    bool loop_added_flag_vis_ = false; //for vis
+    ///// visualize
+    tf::TransformBroadcaster broadcaster_;
+    pcl::PointCloud<pcl::PointXYZ> odoms_, corrected_odoms_;
+    nav_msgs::Path odom_path_, corrected_path_;
+    bool global_map_vis_switch_ = true;
+    ///// results
+    bool save_map_bag_ = false, save_map_pcd_ = false, save_in_kitti_format_ = false;
+    ///// ros
+    ros::NodeHandle nh_;
+    ros::Publisher corrected_odom_pub_, corrected_path_pub_, odom_pub_, path_pub_;
+    ros::Publisher corrected_current_pcd_pub_, corrected_pcd_map_pub_, loop_detection_pub_;
+    ros::Publisher realtime_pose_pub_;
+    ros::Publisher debug_src_pub_, debug_dst_pub_, debug_coarse_aligned_pub_, debug_fine_aligned_pub_;
+    ros::Timer loop_timer_, vis_timer_;
+    // odom, pcd sync, and save flag subscribers
+    std::shared_ptr<message_filters::Synchronizer<odom_pcd_sync_pol>> sub_odom_pcd_sync_ = nullptr;
+    std::shared_ptr<message_filters::Subscriber<nav_msgs::Odometry>> sub_odom_ = nullptr;
+    std::shared_ptr<message_filters::Subscriber<sensor_msgs::PointCloud2>> sub_pcd_ = nullptr;
+    ros::Subscriber sub_save_flag_;
+    ///// Loop closure
+    LoopClosureConfig lc_config_;
+    std::unique_ptr<LoopClosure> loop_closure_;
```

Comment:
I think all things should be set as `std::unique_ptr` because we don't share those instances with other classes. I leverage the explanations of ChatGPT in Korean for better understanding.

```
unique_ptr은 객체에 대한 독점적 소유권을 제공하며, 리소스가 단일 소유자에 의해 관리되어야 할 때 사용됩니다.
shared_ptr은 객체에 대한 공유 소유권을 제공하며, 여러 소유자가 리소스를 공유해야 할 때 사용됩니다.
```

Reply:
I haven’t thought deeply about it, and since the object is not used elsewhere, I used unique_ptr to handle it uniquely. However, it seems better to use `shared_ptr` for consistency.